WMIAdventure/backend/WMIAdventure_backend/IngameUsers/businesslogic/upgrading_cards.py

```diff
@@ -0,0 +1,74 @@
+from typing import Optional
+
+from django.db import transaction
+from django.db.models import Max
+
+from IngameUsers.businesslogic.exceptions import CannotUpgradeCardException
+from IngameUsers.models import UserCard, UserStats
+from cards.models import Card
+
+
+def upgrade_card(user_card: UserCard) -> UserCard:
+    """
+    Upgrades given user's card.
+
+    :param user_card:
+    :return: Upgraded UserCard.
+    :raises CannotUpgradeCardException: If upgrading criteria are not met. (see wiki for more info about those criteria)
+    """
+
+    _check_next_level_cost(user_card)
+    upgraded_card = _get_next_level(user_card.card)
+    if upgraded_card is None:
+        raise CannotUpgradeCardException("Card has no next level")
+
+    # Lock user's stats and user's card for transaction to prevent race conditions
```

Comment:
Super, że o tym pomyślałeś, przyda się jakbyśmy kiedyś chcieli zmienić serwer na ASGI (asynchroniczne requesty)

Powinniśmy w sumie od samego początku robić wszystko asynchronicznie, można na kiedyś zrobić jakiś PoC i zweryfikować, czy w ogóle nasza aplikacja jest na to gotowa

Reply:
fajnie by było, trzeba kiedyś to zrobić